firedrake/preconditioners/hiptmair.py

```diff
@@ -0,0 +1,185 @@
+            self.pc.apply(X, Y)
+
+    def applyTranspose(self, pc, X, Y):
+        dm = self._dm
+        with dmhooks.add_hooks(dm, self, appctx=self._ctx_ref):
+            self.pc.applyTranspose(X, Y)
+
+    def view(self, pc, viewer=None):
+        super(TwoLevelPC, self).view(pc, viewer)
+        if hasattr(self, "pc"):
+            viewer.printfASCII("Two level PC\n")
+            self.pc.view(viewer)
+
+
+class HiptmairPC(TwoLevelPC):
+
+    _prefix = "hiptmair_"
+
+    def coarsen(self, pc):
+        from firedrake_citations import Citations
+        from firedrake import FunctionSpace, TestFunction, TrialFunction
+        from firedrake.interpolation import Interpolator
+        from ufl.algorithms.ad import expand_derivatives
+        from ufl import (FiniteElement, TensorElement, FacetElement,
+                         replace, zero, grad, curl, as_vector)
+
+        Citations().register("Hiptmair1998")
+        appctx = self.get_appctx(pc)
+        V = dmhooks.get_function_space(pc.getDM())
+        ctx = dmhooks.get_appctx(pc.getDM())
+        problem = ctx._problem
+        a = problem.Jp or problem.J
+        bcs = problem.bcs
+
+        mesh = V.mesh()
+        element = V.ufl_element()
+        formdegree = V.finat_element.formdegree
+        if formdegree == 1:
+            dminus = grad
+            cfamily = "Lagrange"
+            G_callback = appctx.get("get_gradient", None)
+        elif formdegree == 2:
+            dminus = curl
+            if V.shape:
+                dminus = lambda u: as_vector([curl(u[k, ...]) for k in range(u.ufl_shape[0])])
+            cfamily = "N1curl" if mesh.ufl_cell().is_simplex() else "NCE"
+            G_callback = appctx.get("get_curl", None)
+        else:
+            raise ValueError("Hiptmair decomposition not available for", element)
+
+        variant = element.variant()
+        degree = element.degree()
+        try:
+            degree = max(degree)
+        except TypeError:
+            pass
+
+        celement = FiniteElement(cfamily, cell=mesh.ufl_cell(), degree=degree, variant=variant)
+        if degree > 1:
+            if not V.finat_element.entity_dofs()[V.finat_element.cell.get_dimension()][0]:
+                celement = FacetElement(celement)
+                # TODO provide statically-condensed form with SLATE
+        if V.shape:
+            celement = TensorElement(celement, shape=V.shape)
+
+        coarse_space = FunctionSpace(mesh, celement)
+        coarse_space_bcs = [bc.reconstruct(V=coarse_space, g=0) for bc in bcs]
+
+        # Get only the zero-th order term of the form
+        beta = replace(expand_derivatives(a), {grad(t): zero(grad(t).ufl_shape) for t in a.arguments()})
+
+        test = TestFunction(coarse_space)
+        trial = TrialFunction(coarse_space)
+        coarse_operator = beta(dminus(test), dminus(trial), coefficients={})
+        if formdegree > 1 and degree > 1:
+            shift = appctx.get("hiptmair_shift", None)
```

Comment:
Should the default be to shift?

Reply:
At low-order with point-Jacobi smoothers in the H(curl) space there is no need to do the shift

Reply:
OK, so the default should be a function of `(low-order && point-jacobi)`?